File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Molecules/TypeBar/MolTypeBar.swift
import Resources
import SwiftUI

public struct MolTypeBar: View {
    @Environment(\.theme) private var theme

    private let text: String
    private let placeholder: String
    private let onTextChange: (String) -> Void
    private let onSubmit: () -> Void
    private let isSubmitEnabled: Bool

    public init(
        text: String,
        placeholder: String,
        onTextChange: @escaping (String) -> Void,
        onSubmit: @escaping () -> Void,
        isSubmitEnabled: Bool
    ) {
        self.text = text
        self.placeholder = placeholder
        self.onTextChange = onTextChange
        self.onSubmit = onSubmit
        self.isSubmitEnabled = isSubmitEnabled
    }

    public var body: some View {
        HStack(spacing: SpacingRoles.Component.compact.rawValue) {
            TextField(
                "",
                text: Binding(get: { text }, set: onTextChange),
                prompt: Text(placeholder).foregroundStyle(theme.colors.onScrimVariant)
            )
            .font(theme.typography.body.p1)
            .foregroundStyle(theme.colors.onScrim)
            .tint(theme.colors.primary)
            .submitLabel(.send)
            .onSubmit(onSubmit)
            Button(action: onSubmit) {
                AtomIcon(
                    icon: Asset.Icons.send,
                    iconSize: .m,
                    contentDescription: L10n.Accessibility.sendMessage,
                    colorRole: submitColorRole
                )
            }
            .buttonStyle(.plain)
            .disabled(!isSubmitEnabled)
        }
        .padding(.leading, theme.padding.elementExpanded)
        .padding(.trailing, theme.padding.elementRelaxed)
        .padding(.vertical, theme.padding.elementRelaxed)
        .frame(minHeight: theme.sizing.interactionHeightDefault)
        .background(theme.colors.scrimContainer, in: RoundedRectangle(cornerRadius: theme.sizing.radiusXL))
    }

    private var submitColorRole: IconColorRole {
        switch isSubmitEnabled {
        case true: .primary
        case false: .onScrimVariant
        }
    }
}

File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/molecules/typebar/MolTypeBar.kt
package com.perrystreet.woof.designsystem.atomic.molecules.typebar

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import com.perrystreet.woof.designsystem.atomic._tokens.sizing.SizingRoles
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.PaddingRoles
import com.perrystreet.woof.designsystem.atomic.atoms.icon.AtomIcon
import com.perrystreet.woof.designsystem.atomic.atoms.icon.roles.IconColorRole
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextColorRole
import com.perrystreet.woof.designsystem.atomic.atoms.text.roles.TextFontRole
import com.perrystreet.woof.designsystem.atomic.atoms.textfield.AtomTextField
import com.perrystreet.woof.designsystem.theme.Theme
import com.perrystreet.woof.resources.R

@Composable
fun MolTypeBar(
    text: String,
    placeholder: String,
    onTextChange: (String) -> Unit,
    onSubmit: () -> Unit,
    modifier: Modifier = Modifier,
    isSubmitEnabled: Boolean = text.isNotBlank(),
) {
    val shape = RoundedCornerShape(Theme.radius.xl)
    val submitColorRole = when (isSubmitEnabled) {
        true -> IconColorRole.Primary
        false -> IconColorRole.OnScrimVariant
    }

    Row(
        modifier = modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = SizingRoles.InteractionHeight.Default.dp)
            .clip(shape)
            .background(Theme.colors.scrimContainer)
            .padding(
                start = PaddingRoles.Element.Expanded.dp,
                end = PaddingRoles.Element.Relaxed.dp,
                top = PaddingRoles.Element.Relaxed.dp,
                bottom = PaddingRoles.Element.Relaxed.dp,
            ),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        AtomTextField(
            text = text,
            placeholder = placeholder,
            onTextChange = onTextChange,
            onSubmit = onSubmit,
            textFontRole = TextFontRole.BodyP1,
            colorRole = TextColorRole.OnScrim,
            placeholderColorRole = TextColorRole.OnScrimVariant,
            modifier = Modifier.weight(1f),
        )
        AtomIcon(
            iconRes = R.drawable.ic_send,
            iconSize = SizingRoles.Icon.M,
            contentDescription = stringResource(R.string.accessibility_send_message),
            colorRole = submitColorRole,
            modifier = Modifier
                .semantics { role = Role.Button }
                .clickable(enabled = isSubmitEnabled, onClick = onSubmit),
        )
    }
}
